CloudEvents Project Security Self-Assessment - Security Pals (#1186)
cncf/tag-security: assessments/projects/cloudevents/self-assessment.md

  * [[Action Item 2] Regular SDK Security
    Audits](#action-item-2-regular-sdk-security-audits)
  * [[Action Item 3] Enforce governance rules throughout SDK
    development](#action-item-3-enforce-governance-rules-throughout-sdk-development)
  * [[Action Item 4] Enhanced Encryption and Data Validation
    Mechanisms](#action-item-4-enhanced-encryption-and-data-validation-mechanisms)
  * [[Action Item 5] Documentation and Knowledge
    Sharing](#action-item-5-documentation-and-knowledge-sharing)
  * [[Action Item 6] Community Engagement and
    Feedback](#action-item-6-community-engagement-and-feedback)
* [Appendix](#appendix)
  * [Known Issues Over Time](#known-issues-over-time)
  * [CII Best Practices](#cii-best-practices)
  * [Case Studies](#case-studies)
  * [Related Projects and Vendors](#related-projects-and-vendors)
* [References](#references)

## Metadata

|   |  |
| -- | -- |
| Software | [CloudEvents Repository](https://github.com/cloudevents/spec)  |
| Security Provider | CloudEvents is not a security provider  |
| CloudEvents Specification Language | Markdown |
| CloudEvents SDK Languages | Python, C#/.NET, Go, Java, Javascript, PHP, PowerShell, Ruby and Rust |

> | Technical Documents | [CloudEvents Documents](https://github.com/cloudevents/spec#cloudevents-documents) |

Reviewer: Please add link to SBOM if the project provides one

Author: Thank you for reviewing the assessment! Unfortunately, CloudEvents does not have a public SBOM, but I added `CloudEvents does not have a public SBOM available` on the table now. Please, let me know if this is fine or if I should make more changes! 

We have also started working on your other comments and will bring more updates soon. Thank you!

logic of the event.

### Actors

#### Event Consumer

The [Event
Consumer](https://github.com/cloudevents/spec/blob/130ba0d183f5e45c1d141f5c1f272cf71d898623/cloudevents/spec.md#consumer)
is the entity or system interested in subscribing to different events. Receiving
Events from the producer will trigger further action that is up to the consumer.
A "consumer" receives the event and acts upon it, which might lead to the
occurrence of new events.

#### Event Producer

The [Event
Producer](https://github.com/cloudevents/spec/blob/130ba0d183f5e45c1d141f5c1f272cf71d898623/cloudevents/spec.md#producer)
is the entity or system that produces the "events". They are responsible for
wrapping event data in the CloudEvents specification.

#### Event Mediator or Intermediary

Depending on the architecture of the system, the [Event
Mediator](https://github.com/cloudevents/spec/blob/130ba0d183f5e45c1d141f5c1f272cf71d898623/cloudevents/spec.md#intermediary)
is the entity or system that is responsible for the distribution, processing and
routing of events to consumers. The event broker ensures reliable delivery and
may enforce security policies.

> 

Reviewer: Does this assessment cover the SDKs offered by CloudEvents team or the scope is limited to the spec? 
If its the former, is the actor/actions updated according to the implementation?

Author: Thank you for noticing that. It covers both the SDK and the Specification, so we added more actors now to cover the SDK part. Please, let me know if I should make more changes.

Context attributes should not carry or represent sensitive information.
CloudEvent producers, consumers, and intermediaries can inspect and log context
attributes.
  
### Security Relevant Components

#### Transport Security

Although CloudEvents does not prescribe specific transport security mechanisms,
it is typically conveyed over secure protocols such as HTTPS, ensuring integrity
and confidentiality.

#### Event Source Authentication

The `source` attribute within a CloudEvent provides context for the event
occurrence, establishing reliable and secure source identification.
  
#### Data Schema Verification

The `data` attribute in a CloudEvent contains the actual event data, and its
schema can be defined and validated for consistency and accuracy. This
verification process helps to prevent issues arising from malformed or
unexpected data.

## Project compliance

As of the latest security assessment, CloudEvents does not explicitly document
compliance with specific security standards such as PCI-DSS, COBIT, ISO, GDPR,
etc. Current efforts are focused on evaluating compliance with these standards
and ensuring that CloudEvents adheres to industry best practices in security and

> privacy.

Reviewer 2: Thanks for the documentation, can you remove the 2nd sentence as it can be misconstrued.

Author: Thank you for reviewing the assessment. Sounds good, I removed the sentence now.

  * [CNCF Slack workspace](http://slack.cncf.io/)
  * [CNCF Meeting Calendar](https://www.cncf.io/community/calendar/)
* Outgoing communications channels:
  * [CloudEvents CNCF
    Subscription](https://lists.cncf.io/g/cncf-cloudevents-sdk)

### Ecosystem

CloudEvents is integrated with different cloud-native technologies and services.
Its goal is to create interoperability in event-driven architecture.

CloudEvents has many
[adopters](https://cloudevents.io/#:~:text=CloudEvents%20Adopters), including:

* Adobe I/O Events
* Azure Event Grid
* European Commission
* Google Cloud Eventarc
* IBM Cloud Code Engine
* Oracle Cloud
* VMware Event Broker Appliance

## Security Issue Resolution

### CloudEvents Specification

To report an issue, or to a new idea, contributors can open a new issue on the
GitHub repository. However, they should check if a similar issue already exists.

CloudEvents also provides a dedicated [email
address](cncf-cloudevents-security@lists.cncf.io) and a [Security Mailing
List](https://lists.cncf.io/g/cncf-cloudevents-security/topics) for reporting
security concerns related to the specification or the SDKs.

### CloudEvents SDKs

> #### Responsible Disclosures Process

Reviewer 2: Provide links to the Responsible Disclosures Process and Vulnerability Response Process documented in CloudEvents. Additionally, update the content in more details.

Author: Thank you for the comment! I added new links to documents CloudEvents has that mention these processes. Unfortunately, they don't have a document outlining these processes in further detail, so I also added a new action item on `Improve the documentation on the Security Response Processes`. I also messaged the CloudEvents team now to ask if they have more details on these processes that I could add to the assessment. I will update it further when they give me more information. Please, let me know if I should make more changes. Thank you!

  data to disrupt services that consume CloudEvents.
* **An External Attacker with valid access:** Could potentially inject false
  events to lead to incorrect actions being taken by consumers.
* **An Internal Attacker:** Might be able to alter event definitions or
  intercept event data in transit within systems that utilize CloudEvents.
* **A Malicious Internal User:** Could misuse the event system to trigger
  unauthorized actions or services.

### Theoretical Threats

* **Deployment Architecture (pod and namespace configuration):** Not applicable
  to CloudEvents as a specification.

* **Networking (internal and external):** Networking controls are typically
  managed by the underlying infrastructure and transport protocols used for
  sending and receiving CloudEvents.

* **Cryptography:** CloudEvents does not provide encryption methods or
  cryptographic implementations, all implementations are left up to the users'
  choice.

* **Audit and logging:** This is typically handled by the platforms and services
  that implement CloudEvents.

### Potential threats

#### The SDK management teams may implement vulnerabilities while implementing SDKs

> Trail Of Bits was able to identify 7 different security concerns regarding the

Reviewer: This seems to be the work outside of the scope of security pals activity. I would recommend introducing it in a few sentences and providing its reference instead of copying the same content here.

Author: Thank you for the feedback. I shortened this sentence and added a reference link there. Please, let me know if I should make more changes there.

newer ones, are susceptible to security vulnerabilities. While governance rules
play a crucial role in ensuring continuous patching and monitoring, some SDKs
have experienced vulnerabilities in the past, highlighting an ongoing concern.

During a [team meeting on November 30, 2023](https://youtu.be/2OZPTQOqFEw), a
CloudEvents administrator also mentioned plans to introduce [bots for regularly
assessing the health of each
SDK](https://github.com/cloudevents/spec/issues/1235). This proactive measure
aims to address potential security concerns that may arise in the future,
mitigating both technical and managerial aspects across various SDKs.

### Recommendations

As mentioned by CloudEvents, the maintainers of SDKs must release security
patches and fix issues frequently for the project to be marked as active. It is
essential for the maintainers of each SDK to frequently triage and release
patches to the security issues found, which can solve problems such as the ones
found by Trail Of Bits. Moreover, it is important for users to frequently check
and install updates to their CloudEvents SDK installation.

## Action Items

> ### [Action Item 1] Setup Snyk for SDKs

Reviewer: Kindly keep the recommendations to reflect the task instead of suggesting particular tool. In this case, the task is to establish SAST, SCA and/or secret scanning services. The project team is free to choose the tool of their choice to accomplish this. 

In addition, pointer to associated task/process documentation could be useful, however I would suggest removing the steps to accomplish it embedded in this document.

Author: Thank you for the comment! I changed the comment to make it broader now. I also linked this to a GitHub issue with the task/process documentation and removed the steps from the documentation.

the entire software development lifecycle, encompassing key stages like
building, rigorous testing, and seamless code deployment. Given the thorough of
the CloudEvents Governance rules, integrating a CI/CD pipeline would both ensure
adherence to these rules and fortify security measures, promoting a more robust
environment for the development of the SDKs. This would lower the amount of
human mistakes in this process and, consequently, the chances of new
vulnerabilities being created.

#### [Action Item 3] Steps

Implementing a CI/CD pipeline for CloudEvents involves using Git for source code
management and selecting a CI/CD platform like Jenkins or GitHub Actions. The
pipeline should include a build stage to compile code and create artifacts, and
a test stage for unit and integration tests, with a focus on validating
CloudEvents generation and consumption. Artifacts should be stored in container
registries or repositories, and the deployment stage completes the process by
deploying to the target environment. This automated pipeline ensures efficient
and reliable development, testing, and deployment of applications using
CloudEvents.

> ### [Action Item 4] Enhanced Encryption and Data Validation Mechanisms

Reviewer: This title is not relevant to the content, as it implies project needs to implement enhanced encryption and data validation mechanisms. Something on the lines of document project scope and user responsibilities might a better fit.

Author: Thank you for noticing that, I changed the title to `Emphasize goals, non-goals and user responsibilities`.

Create comprehensive and accessible documentation that covers security best
practices, identified vulnerabilities, resolved issues, public Software Bill of
Materials, and strategies for mitigation of vulnerabilities.

### [Action Item 6] Community Engagement and Feedback

#### [Action Item 6] Objective

Enhance engagement with the CloudEvents user community for feedback and insights
on security.

#### [Action Item 6] Steps

Establish channels for consistent feedback, administer community surveys, and
arrange Q&A sessions. Evaluate GitHub issues, engage in discussions with
authors, formulate mitigation or solution strategies, and execute them.
Prioritize and resolve security vulnerabilities identified by users through
proper triage processes.

## Appendix

### Known Issues Over Time

#### Issues found by Trail of Bits

The main security assessment on CloudEvents was a [Security Audit performed by
Trail of
Bits](https://github.com/cloudevents/spec/blob/main/docs/CE-SecurityAudit-2022-10.pdf)
released on October 26, 2022. These were mainly concerning the different
CloudEvents SDK, not the specification. All of the security issues found by
Trail of Bits have already been addressed.

> Below are listed the findings by Trail Of Bits with their descriptions.

Reviewer: This seems to be the work outside of the scope of security pals activity. I would recommend introducing it in a few sentences and providing its reference instead of copying the same content here.

Author: Thank you for the feedback. We removed this part and replaced it with a summary and reference.

function parses a machine-dependent integer (int64 for 64-bit targets). In some
code instances, the result from strconv.Atoi is later converted to a smaller
type (int16 or int32), risking overflow with specific inputs.

##### [Go SDK] ReadHeaderTimeout not configured

* **Severity:** Informational
* **Difficulty:** Low
* **Type:** Denial of Service
* **Finding ID:** TOB-CE-6
* **Target:** Go SDK

The [Go SDK](https://github.com/cloudevents/sdk-go) http.server API offers four
timeouts, including ReadHeaderTimeout. Failure to set a value for this timeout
makes the listener instance susceptible to Slowloris DoS attacks.

##### [CSharp SDK] Outdated Vulnerable Dependencies

* **Severity:** Undetermined
* **Difficulty:** Low
* **Type:** Patching
* **Finding ID:** TOB-CE-7
* **Target:** CSharp SDK

The [CSharp SDK](https://github.com/cloudevents/sdk-csharp) has multiple
outdated dependencies with known vulnerabilities. Using the open-source Snyk
tool, each module was automatically audited. Due to time constraints and ease of
remediation, a manual review of exploitability within the SDK's context was
skipped.

### CII Best Practices

> 

Reviewer 2: Add explicitly that the project doesn't have one CII best practices badge. You also consider one of the action items to improve alignment with CII best practices and to get one.

Author: Thank you for the comment. I found that CloudEvents received a `passing` [badge in 2022](https://www.bestpractices.dev/en/projects/6770), so I just added it to the assessment now. I also added a new action item on `Obtain a Silver or Gold CII Best Practices Badge`. Please, let me know if this is good or if I should make more changes!

Reviewer 2: Yup, the changes good.

various sectors. Below are a couple of enhanced case studies demonstrating its
applicability.

#### Healthcare Data Management

CloudEvents plays a pivotal role in a healthcare system for managing patient
data. In this scenario, it is used to track and route events such as patient
admissions, lab results, and medication orders. By standardizing event formats,
CloudEvents ensures seamless integration between different hospital systems,
including Electronic Health Records (EHRs) and Laboratory Information Management
Systems (LIMS), enhancing patient care and operational efficiency.

#### Smart City Infrastructure Monitoring

In a smart city environment, CloudEvents is employed to aggregate and analyze
data from a multitude of sensors and IoT devices across the city. This includes
traffic flow sensors, public transport updates, and utility usage meters. By
utilizing CloudEvents for real-time data processing and event routing, city
administrators can make informed decisions about traffic management, public
transport schedules, and resource allocation, leading to improved city services
and resident satisfaction.

> ### Related Projects and Vendors

Reviewer: This section is intended to introduce related projects, and state how it is different from the current project in scope. I see the former, however the latter is absent. Please add a few sentences on how these projects are different from CloudEvents.

Author: Thank you for the feedback. I added a paragraph under each related project to compare them with CloudEvents. Please, let me know if I should add more details there.

# CloudEvents self-assessment

September 26, 2023

Authors: Igor Rodrigues (@Igor8mr), Matthew Gong (@MatthewZGong), Kushal Kothari
(@Kushal-kothari) and Devyani Bairagya (@devyani-14).

Contributors/Reviewers: Pranava Kumar Vemula (@Rana-KV), Doug Davis (@duglin),
Eddie Knight (@eddie-knight) and Ragashree (@ragashreeshekar).

> This document elaborates and explores the design goals for
> [CloudEvents](https://cloudevents.io/) as well as a security assessment of the
> software. It was written in collaboration with the CloudEvents team. All
> communications with CloudEvents were recorded on a [separate
> document](CE-maintainers-communications.md).

Reviewer 3: I like that you put this here! But since we removed the comms file, we should restructure the paragraph.

```suggestion
This document was written in collaboration with the CloudEvents team to elaborate on and explores the design goals for
[CloudEvents](https://cloudevents.io/) through a limited security assessment of the software.
```

Author: Thank you, I committed the suggestion.

  * [Theoretical Threats](#theoretical-threats)
  * [Potential threats](#potential-threats)
  * [Conclusion](#conclusion)
  * [Recommendations](#recommendations)
* [Action Items](#action-items)
  * [[Action Item 1] Setup a system for automatically reviewing
    SDKs](#action-item-1-setup-a-system-for-automatically-reviewing-sdks)
  * [[Action Item 2] Regular SDK Security
    Audits](#action-item-2-regular-sdk-security-audits)
  * [[Action Item 3] Enforce governance rules throughout SDK
    development](#action-item-3-enforce-governance-rules-throughout-sdk-development)
  * [[Action Item 4] Obtain a Silver or Gold CII Best Practices
    Badge](#action-item-4-obtain-a-silver-or-gold-cii-best-practices-badge)
  * [[Action Item 5] Improve the documentation on the Security Response
    Processes](#action-item-5-improve-the-documentation-on-the-security-response-processes)
  * [[Action Item 6] Emphasize goals, non-goals and user
    responsibilities](#action-item-6-emphasize-goals-non-goals-and-user-responsibilities)
  * [[Action Item 7] Documentation and Knowledge
    Sharing](#action-item-7-documentation-and-knowledge-sharing)
  * [[Action Item 8] Community Engagement and

>     Feedback](#action-item-8-community-engagement-and-feedback)

Reviewer 3: Following discussions from the review team, we are asking that Security Pals please remove any recommendations from the scope of this document (unless they were written and included under guidance from the maintainer team). You may optionally communicate Security Pals recommendations for projects in a more accessible location, such as in a project community channel or GitHub Issue.

Author: Hi @eddie-knight, thank you for the feedback on this. I removed all the action items, except the `Setup a system for automatically reviewing SDKs`, which is something we discussed with the maintainer team, as you mentioned not to remove those. They are also already [working on implementing this](https://github.com/cloudevents/spec/issues/1235) and we provided some [guidance to help them get started](https://github.com/cloudevents/spec/issues/1235#issuecomment-1841532959).  I also moved this action item to the appendix now as you mentioned.

Please, let me know if I should still change this section more, or remove this last action item. Thank you again!

* [Appendix](#appendix)
  * [Known Issues Over Time](#known-issues-over-time)
  * [CII Best Practices](#cii-best-practices)
  * [Case Studies](#case-studies)
  * [Related Projects and Vendors](#related-projects-and-vendors)
* [References](#references)

## Metadata

|   |  |
| -- | -- |
| Software | [CloudEvents Repository](https://github.com/cloudevents/spec)  |
| Security Provider | CloudEvents is not a security provider  |
| CloudEvents Specification Language | Markdown |
| CloudEvents SDK Languages | Python, C#/.NET, Go, Java, Javascript, PHP, PowerShell, Ruby and Rust |

> | SBOM | CloudEvents does not have a public SBOM available |

Author: Thank you for the suggestion! I committed it.

The [Apex Event Specification Guide
(ESG)](https://insights.eventscouncil.org/Portals/0/APEX_Event_Specifications_Guide.pdf)
is a written document that contains all the details of an event. The ESG is used
by event organizers to communicate information to venues and suppliers. The ESG
is a three-part template that includes: Narrative, Schedule, and Function
orders.

This is an older document used across many engineering fields, which is not
restricted to computer science, making it different from the other examples.
However, it is an example of event specification being widely used in industry,
including more physical areas such as mechanical engineering and factories.

## References

* [CloudEvents Website](https://cloudevents.io/)
* [CloudEvents GitHub](https://github.com/cloudevents)
* [CloudEvents Security Assessment by Trail of
  Bits](https://github.com/cloudevents/spec/blob/main/docs/CE-SecurityAudit-2022-10.pdf)
* [Security Assessment Guide](https://github.com/Rana-KV/ISP)
* [Sample Security
  Assessment](https://github.com/Rana-KV/tag-security/blob/main/assessments/projects/karmada/self-assessment.md#threat-modeling-with-stride)
* [Open and Secure

>   Book](https://github.com/cncf/tag-security/blob/main/assessments/Open_and_Secure.pdf)

Author: Thank you for your comment, I merged it to the appendix now.

* Simplify event declaration and delivery across services, platforms and
  systems.
* Define interoperability of event systems that allow services to produce or
  consume events, where the producer and consumer can be developed and deployed
  independently.
* Establish a structured framework for the exchange of event data among services
  by standardizing the location for storing and retrieving common metadata
  related to events.
* Provide specifications for serializing events in different formats and
  protocols.

### Non-goals

The following are considered as [non-goals of
CloudEvents](https://github.com/cloudevents/spec/blob/main/cloudevents/primer.md#non-goals):

* Function build and invocation process.
* Language-specific runtime APIs.
* Selecting a single identity/access control system.
* Inclusion of protocol-level routing information.
* Event persistence processes.
* Mechanism for Authorization, Data Integrity and Confidentiality.
* Change the implementation details of underlying communication protocols.
* Define the processing logic for events within different applications and
  systems.

## Self-assessment use

> This self-assessment is created by the CloudEvents team to perform an internal

Reviewer 4: Can we modify this to more accurately state who created it? It's fair to add that the CE team has reviewed it but the CE team did not create it. There might be other spots in here that might need similar tweaking.

I'd also like to have something added about how the "source of truth" for all things CE related can be found in the CE project repos and this is a point-in-time summary of the project.

Author: Thank you for the feedback on this. That makes sense, I changed it now to be more precise on who created and reviewed it.

> This security assessment was created by [Igor Rodrigues](https://github.com/Igor8mr), [Matthew Gong](https://github.com/MatthewZGong), [Kushal Kothari](https://github.com/Kushal-kothari) and [Devyani Bairagya](https://github.com/devyani-14) to perform an internal analysis of the CloudEvents project. The document was also reviewed by CloudEvents maintainers and members of the [CNCF Security Technical Advisory Group (TAG)](https://github.com/cncf/tag-security). It is meant to provide a current summary of the project and its security-related aspects. It is not intended to provide a security audit of CloudEvents, or function as an independent assessment or attestation of CloudEvents's security health. For the latest state of the project please check the [CloudEvents GitHub repository](https://github.com/cloudevents/spec).

I also added the paragraph below to the beginning of the document to mention where we extracted this information from:

> The information contained in this assessment was extracted from the [CloudEvents GitHub Repository](https://github.com/cloudevents/spec), and other channels maintained by the project, such as its [website](https://cloudevents.io/) and [Slack workspace](http://slack.cncf.io/).

Please, let me know if you want us to change these more or change other parts of the documents. Thank you for the help!

> analysis of the project's security. It is not intended to provide a security

Reviewer 4: I'd change this to something like:
This assessment was created by ..., and has been reviewed by the CloudEvents team. It is meant to provide a point-in-time summary the state of the project and summarize its security related aspects. It is not intended to provide a security audit of CloudEvents, or function as an independent assessment or attestation of CloudEvents' security health. For the latest state of the project please see the CloudEvents github [repository](https://github.com/cloudevents/spec).

Author: Thank you for the suggestion, I incorporated it into the assessment with some modifications. Please, let me know if you want to make more changes to it.

> # CloudEvents self-assessment
> 
> November 26, 2023
> 

Reviewer 5: ```suggestion
# CloudEvents self-assessment

This assessment was created by community members as part of the [Security Pals](https://github.com/cncf/tag-security/issues/1102) process and is currently pending changes from the maintainer team.

November 26, 2023

```
@ragashreeshekar
